### PR TITLE
[`pylint`] - Implement `too-many-attributes rule` (`PLR0902`)

### PR DESCRIPTION
**Pull request by liushuyu**
## Summary

Adds too-many-attributes rule (PLR0902) from Pylint. This implementation has a similar limitation as the Pylint one, where dynamically generated class attributes (e.g. through `setattr` or `eval`) are not considered.

## Test Plan

A new test has been added and can be executed using `cargo test`.

### CODE REVIEW

**Review by MichaReiser — COMMENTED** [review]
Thanks for contributing this rule.

Code: There are two cases where the rule currently misses field declarations that should be supported. 

An open question for us as maintainers is whether we want to merge the rule today. The problem we're facing is that many users use `--select ALL` and consider it as Ruff's "recommended" rule set. I see that this rule is useful when used in a mindful way, but it could guide less aware users in the direction that having classes with many attributes is always bad and is generally considered bad practice. 

We hope to solve the `ALL` problem by recategorizing our rules so that more opinionated rules require explicit opt-in.  But we may need to wait to add new rules till then.

**Comment by MichaReiser** [comment]
> This lint is current in the experimental state, so it won't be enabled by default ...?

That's correct. But the intent would be to stabilize the rule eventually (ideally one or two minor releases later).

**Comment by liushuyu** [comment]
> > This lint is current in the experimental state, so it won't be enabled by default ...?
> 
> That's correct. But the intent would be to stabilize the rule eventually (ideally one or two minor releases later).

Understood. So this pull request would be put on the backburner and wait until a better grouping mechanism is added?

**Comment by MichaReiser** [comment]
I discussed this internally and @charliermarsh is fine with adding new pylint rules, even if they are very opinionated.

File: crates/ruff_linter/resources/test/fixtures/pylint/too_many_attributes.py
Comment (MichaReiser) [inline crates/ruff_linter/resources/test/fixtures/pylint/too_many_attributes.py:5]: Nice catch. I wasn't aware that this is allowed too

File: crates/ruff_linter/src/checkers/ast/analyze/deferred_scopes.rs
Comment (MichaReiser) [inline crates/ruff_linter/src/checkers/ast/analyze/deferred_scopes.rs:316]: Should the rule name be `too-many-instance-attributes` to match [pylint's](https://pylint.readthedocs.io/en/stable/user_guide/messages/refactor/too-many-instance-attributes.html) name?

File: crates/ruff_linter/src/rules/pylint/rules/too_many_attributes.rs
Comment (MichaReiser) [inline crates/ruff_linter/src/rules/pylint/rules/too_many_attributes.rs:66]: I suspect that we don't need the `scope` here, because it's given that we're inside of a class, because we get the `class_def`

**Review by MichaReiser — APPROVED** [review]
This looks good to me but I want that @charliermarsh takes a look because I wonder if we extract the variables already as part of the semantic model ( I tried `scope.bindings` but this doesn't give me all fields).

File: crates/ruff_linter/resources/test/fixtures/pylint/too_many_attributes.py
Comment (MichaReiser) [inline crates/ruff_linter/resources/test/fixtures/pylint/too_many_attributes.py:20]: Can we add a few more tests:

* a test for a class that has exactly 7 attributes
* a test for a class that assigns to attributes in method other than `__init__`
* a test for a class that has fewer than 7 attributes
* a test for a class that assigns the same variable multiple times (e.g. defines the attributes and assigns in `__init__`

File: crates/ruff_linter/src/rules/pylint/rules/too_many_attributes.rs
Comment (liushuyu) [reply crates/ruff_linter/src/rules/pylint/rules/too_many_attributes.rs]: Addressed.

File: crates/ruff_linter/src/checkers/ast/analyze/deferred_scopes.rs
Comment (liushuyu) [reply crates/ruff_linter/src/checkers/ast/analyze/deferred_scopes.rs]: Done.

**Comment by liushuyu** [comment]
Comments addressed and rebased against the latest main branch.

**Comment by MichaReiser** [comment]
I think this is the same as https://github.com/astral-sh/ruff/pull/10431. We need to decide which of the two PRs we want to move forward.

**Comment by diceroll123** [comment]
> I think this is the same as #10431. We need to decide which of the two PRs we want to move forward.

I hadn't noticed this PR existed before making my own, I concede to this one!

**Comment by MichaReiser** [comment]
Thank you for contributing to this new rule, and I am sorry that it took so long to decide.

 We agree that the rule itself is useful, and we're open to merging the rule once #1774 is complete, but we don't want to merge the rule today. The reasons are:

* The rule is very opinionated. It intentionally restricts Python functionality by disallowing classes with more than X attributes. This is fine, but we currently have no good model to group such opinionated rules, and we don't want users using `ALL` or selecting pylint to enable this rule by default because we believe that would be misleading and not all users have the required knowledge to know that this rule might just not be for them. 
* The rule measures complexity similar to the mc-cabe rule. It is different in that it measures the complexity of a structure and not a control flow complexity. However, we must figure out how we want different metrics to measure complexity. Should these be separate rules or a single, configurable rule? 

I'm sorry that it took us so long to come to that conclusion and that we haven't flagged this on the pylint issue. I plan to go through the Pylint rules and mark the rules that are unlikely to be accepted today, but I first have to catch up on open linter PRs. 

Thank you again for contributing and I hope we can merge this rule in the future, once we figured out the right place for it.

**Pull request closed without merging**
